pkg/util/topology.go

```diff
@@ -0,0 +1,240 @@
+/*
+Copyright 2020 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+const (
+	keySeparator   rune   = '/'
+	labelSeparator string = ";"
+)
+
+// GetDomainsFromLabels reads passed in domain labels from the k8s node
+// and determines their values. It returns the domain name and their values, sans
+// the label prefix, as a map for further use. expects domainLabels in arg to be
+// in the format "[prefix/]<name>;[prefix/]<name>;...",
+func GetDomainsFromLabels(domainLabels, nodeName string) (map[string]string, error) {
+	var client *k8s.Clientset
+
+	if domainLabels == "" {
+		return nil, nil
+	}
+
+	// Convert passed in labels to a map, and check for uniqueness
+	labelsToRead := strings.SplitN(domainLabels, labelSeparator, -1)
+	klog.Infof("passed in node labels for processing : %+v", labelsToRead)
+
+	labelsIn := make(map[string]bool)
+	labelCount := 0
+	for _, label := range labelsToRead {
+		// as we read the labels from k8s, and check for missing labels,
+		// no label conformance checks here
+		if _, ok := labelsIn[label]; ok {
+			return nil, fmt.Errorf("duplicate label (%s) found in domain labels", label)
+		}
+
+		labelsIn[label] = true
+		labelCount++
+	}
+
+	// Create a k8s client to read node labels
+	client = NewK8sClient()
+	node, err := client.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
```

Comment:
@ShyamsundarR this looks to be problematic to me. ooc, have we tested this PR in a setup?  reading `node` object need special permissions , Isnt it ?

Reply:
There are a few ways to deal with this, for example to pass in the domain label values to the nodeplugins or create subset deployments per lowest domain with static labels. This is discussed [here](https://github.com/ceph/ceph-csi/pull/760/files#r360098011)

The manner of handling reading node labels seems the best when using daemon sets, as we need a nodeplugin running on all nodes. This does require the required roles and permissions.

Cloud providers reach into their API endpoints ([example](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/7c1a7afee2c79bdfe3f934ae8c4e839addc55720/pkg/cloud/metadata.go#L81-L106)) to get the node information, as these are used to both set the node labels (k8s) and also return CSI node domains. Further, they have access to the required credentials to make the said API calls. We lack such abilities, as we do not decide on the node labels. as these are setup by the cluster admin or the cloud provider.

In the face of the above limitations requiring additional permissions is required.